File: 0872-leaf-similar-trees/0872-leaf-similar-trees.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    
    
    public void getLeaves(TreeNode root1, List<Integer> leaves)
    {
        if(root1 == null)
            return;
        
        if(root1.left == null && root1.right == null)
            leaves.add(root1.val);
        
        getLeaves(root1.left, leaves);
        getLeaves(root1.right, leaves);
            
    }
    
    
    public boolean same(List<Integer> l1, List<Integer> l2)
    {
        int n1 = l1.size();
        int n2 = l2.size();
        if(n1 != n2)
            return false;
        
        for(int i = 0; i<=n1-1; i++)
        {
            if(l1.get(i) != l2.get(i))
                return false;
        }
        return true;
    }
    
    
    public boolean leafSimilar(TreeNode root1, TreeNode root2) {
        List<Integer> l1 = new ArrayList<>();
        List<Integer> l2 = new ArrayList<>();
        
        getLeaves(root1, l1);
        getLeaves(root2, l2);
        
        
        return same(l1, l2);
    }
}

File: 0872-leaf-similar-trees/0872-leaf-similar-trees.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root1
 * @param {TreeNode} root2
 * @return {boolean}
 */

function getLeaves(root, leaves) {
    if (root == null) return;

    if (root.left == null && root.right == null) {
        leaves.push(root.val);
    }

    getLeaves(root.left, leaves);
    getLeaves(root.right, leaves);
}

function same(l1, l2) {
    if (l1.length != l2.length) {
        return false;
    }

    for (let i = 0; i <= l1.length - 1; i++) {
        if (l1[i] != l2[i]) {
            return false;
        }
    }

    return true;
}

function leafSimilar(root1, root2) {
    const l1 = [];
    const l2 = [];

    getLeaves(root1, l1);
    getLeaves(root2, l2);

    return same(l1, l2);
}
